node_modules/oae-tenants/lib/api.js

```diff
@@ -494,6 +534,64 @@ var _setLandingPageBlockAttribute = function(ctx, block, blockName, attributeNam
 ///////////////
 
 /**
+ * Ensure that the given email domain does not conflict with any existing tenant domains. If this
+ * validation is part of an update, the update tenant alias should be specified so that any conflict
+ * in email domain with the tenant being updated can be ignored. Any validation errors will result
+ * in the errors being added to the provided `validator`, therefore this does not `return` or
+ * `throw` anything.
+ *
+ * @param  {Validator}  validator               The Validator object
+ * @param  {String}     emailDomain             The email domain expression to verify
+ * @param  {String}     [updateTenantAlias]     The alias of the tenant being updated, if any
+ * @api private
+ */
+var _validateEmailDomain = function(validator, emailDomain, updateTenantAlias) {
+    var matchingEmailDomain = _.chain(getTenants())
+        .filter(function(tenant) {
+            // If a tenant is being updated, don't try and match against its own existing email
+            // domain
+            return (tenant.alias !== updateTenantAlias);
+        })
+        .pluck('emailDomain')
+        .compact()
+        .filter(function(emailDomainExpr) {
+            // There is a conflict if there is an existing expression that matches with the
+            // expression, or vice versa
+            return (_matchesEmailDomain(emailDomain, emailDomainExpr) ||
+                _matchesEmailDomain(emailDomainExpr, emailDomain));
+        })
+        .first()
+        .value();
+
+    validator.check(matchingEmailDomain, {
+        'code': 400,
+        'msg': util.format('The email domain "%s" conflicts with existing email domain: %s', emailDomain, matchingEmailDomain)
+    }).isNull();
+};
+
+/**
+ * Determine if the given email domain expression matches the given email domain
+ *
+ * @param  {String}     emailDomain         The email domain to check
+ * @param  {String}     emailDomainExpr     The email domain expression (e.g., *.cam.ac.uk)
```

Comment:
Could we try and "parse" the email domains and look for suffix matches? For example,
we configure `cam.ac.uk` as the Cambridge tenant's email domain and we want to check: `simon@admin.cam.ac.uk`, we explode all the email domain suffices:

1. uk
2. ac.uk
3. cam.ac.uk
4. admin.cam.ac.uk

and do _.contains(emailDomainSuffixes, emailDomain);

Something like this:
```
var _ = require('underscore');

var getEmailSuffices = function(emailDomain) {
    var splittedEmailDomain = emailDomain.split('.');
    var emailSuffices = [];
    for (var i = 0; i < splittedEmailDomain.length; i++) {
       var emailSuffix = [];
       for (var j = i; j < splittedEmailDomain.length; j++) {
           emailSuffix.push(splittedEmailDomain[j]);
       }
       emailSuffices.push(emailSuffix.join('.'));
    }
    return emailSuffices;
};

var checkEmail = function(configuredEmailDomain, email) {
    var emailSuffices = getEmailSuffices(email.split('@')[1]);
    return _.contains(emailSuffices, configuredEmailDomain);
};

var test = function(email) {
    var configuredEmailDomain = 'cam.ac.uk';
    var containsDomain = checkEmail(configuredEmailDomain, email);
    console.log('%s contains %s: %s', email, configuredEmailDomain, containsDomain);
};

test('gaeremycks@gmail.com');
test('sg555@cam.ac.uk');
test('sg555@unicam.ac.uk');
test('support@uis.cam.ac.uk');
test('me@admin.foo.bar.cam.ac.uk');
```